### PR TITLE
Refactor homography operations from kornia-3d and kornia-apriltag

### PR DESCRIPTION
**Pull request by Nyx128**
## 📝 Description

**Fixes/Relates to:** #724 

## 🛠️ Changes Made
- [x] move all homography computation to `kornia-algebra`
- [x] port over any tests from `kornia-3d` and `kornia-apriltag` into the new `homography` module in `kornia-algebra`
- [x] 

---

## 🧪 How Was This Tested?
- [x] **Unit Tests:** All previous tests related to homography compute are ported as is
- [x] **Performance/Edge Cases:** relevant benchmark details have been provided below, ransac homography is quite a bit faster now as a result due to static matrices.

The benchmarks were performed on the following hardware(CPU: AMD Ryzen 7 8845HS, RAM: 16GB DDR5 5600MHz, OS: Win 11(mvsc-rust-toolchain-x86_64))

| Benchmark | Before (median) | After (median) | Change | Speedup |
|---|---|---|---|---|
| `ransac_homography/25` | 7.0729 µs | 3.7740 µs | −3.2989 µs | **1.87× faster (−46.6%)** |
| `ransac_homography/100` | 7.6476 µs | 4.0064 µs | −3.6412 µs | **1.91× faster (−47.6%)** |
| `ransac_homography/500` | 7.7728 µs | 4.4888 µs | −3.2840 µs | **1.73× faster (−42.3%)** |
| `kornia-apriltag` | 5.4052 ms | 5.4445 ms | +0.0393 ms | **0.99× (≈0.7% slower)** |

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [ ] 🟢 **No AI used.**
- [x] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [x] I am assigned to the linked issue (required before PR submission)
- [x] The linked issue has been approved by a maintainer
- [x] This PR strictly implements what the linked issue describes (no scope creep)
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] (Optional) I have attached screenshots/recordings for UI changes.

---

## 💭 Additional Context
The homography functions can be further generalized, but that would be somewhat outside the scope of this refactor.
We could implement functions that take abitrary types and also take solver policy parameters(gauss-elim/svd) to further streamline usage.

### CODE REVIEW

File: crates/kornia-algebra/src/linalg/homography.rs [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+/// # Arguments
+///
+/// * `c` - A 4x4 array where each row contains the coordinates of a correspondence:
+///   [x, y, x', y'] where (x, y) maps to (x', y').
+///
+/// # Returns
+///
+/// An `Option<Mat3F32>` containing the 3x3 homography matrix if successful, or `None` if the matrix is singular.
+pub fn homography_2d_dlt_gauss_elim_f32(c: &[[f32; 4]; 4]) -> Option<Mat3F32> {
+    #[rustfmt::skip]
```
[the remark this answers is not in the document]
Comment (Nyx128) [reply crates/kornia-algebra/src/linalg/homography.rs]: glam based interface has been provided in [23e5839](https://github.com/kornia/kornia-rs/pull/790/commits/23e58396fb95e15b69e32eec9d9e83da322bc513)

File: crates/kornia-algebra/src/linalg/homography.rs [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+    fn test_homography_2d_dlt_gauss_elim_f32() {
+        #[rustfmt::skip]
+        let corr_arr = [
+            [-1.0, -1.0, 27.0,  3.0],
+            [ 1.0, -1.0, 27.0, 27.0],
+            [ 1.0,  1.0,  3.0, 27.0],
+            [-1.0,  1.0,  3.0,  3.0],
+        ];
+
+        let h = homography_2d_dlt_gauss_elim_f32(&corr_arr).unwrap();
```
[the remark this answers is not in the document]
Comment (Nyx128) [reply crates/kornia-algebra/src/linalg/homography.rs]: fixed in [23e5839](https://github.com/kornia/kornia-rs/pull/790/commits/23e58396fb95e15b69e32eec9d9e83da322bc513) now results are used in tests

File: crates/kornia-3d/src/pose/homography.rs [hunk 3]
```diff
@@ -24,44 +23,23 @@ pub fn homography_4pt2d(
...
+        if linalg::det_mat33(homo).abs() < 1e-8 {
+            return Err(HomographyError::SingularMatrix);
+        }
 
-    if linalg::det_mat33(homo).abs() < 1e-8 {
-        return Err(HomographyError::SingularMatrix);
+        Ok(())
+    } else {
+        Err(HomographyError::SingularMatrix)
     }
```
[the remark this answers is not in the document]
Comment (Nyx128) [reply crates/kornia-3d/src/pose/homography.rs]: error handling and homography is completely handled inside kornia-algebra now, kornia-3d will now purely use functions from `kornia-algebra`

File: crates/kornia-algebra/src/linalg/homography.rs [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    // take the 9th row of v_t (which is the last column of v)
+    let v_t = svd.v_t?;
+    let h = v_t.row(8);
+
+    Some(Mat3F64::from_cols_array(&[
+        h[0], h[3], h[6], // col 0
+        h[1], h[4], h[7], // col 1
+        h[2], h[5], h[8], // col 2
+    ]))
+}
```
[the remark this answers is not in the document]
Comment (Nyx128) [reply crates/kornia-algebra/src/linalg/homography.rs]: docs have been edited now with examples for clarity and are overall more clear

File: crates/kornia-3d/src/pose/homography.rs [hunk 5]
```diff
@@ -224,93 +177,3 @@ pub fn decompose_homography(h: &Mat3F64, k1: &Mat3F64, k2: &Mat3F64) -> Vec<(Mat
 
     out
 }
-
-#[cfg(test)]
```
[the remark this answers is not in the document]
Comment (Nyx128) [reply crates/kornia-3d/src/pose/homography.rs:228]: all relevant tests have been ported over, to ensure regression testing

File: crates/kornia-algebra/src/linalg/homography.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Nyx128) [reply crates/kornia-algebra/src/linalg/homography.rs]: this function now does chirality checks and other checks and also returns errors now

**Comment by Nyx128** [comment]
## Changes Made

### 1. Centralized Solvers in `kornia-algebra`
All core homography solvers now reside in kornia-algebra/src/linalg/homography.rs, making them available for all crates.

- `homography_4pt2d_gauss_elim_f32`: High-speed solver for 2D points using Gaussian elimination (replaces `kornia-apriltag` implementation).
- `homography_4pt2d_svd_f64`: Robust 8-point solver for 2D points using SVD.
- `homography_4pt3d_lu_f64`: Solver for 3rd-point correspondences using LU decomposition with cheirality checks.
- `HomographyError`: Centralized error type for singular matrices and cheirality failures, which is ported over from `kornia-3d`

### 2. Modernized API with `glam`
Interfaces have been updated to use `glam` types (`Vec2F32`, `Vec2F64`, `Vec3F64`) for better performance and consistency.
- Solvers now return `Result<Mat3Fxx, HomographyError>`, enforcing robust error propagation.
- Hardcoded magic numbers were replaced with a consistent `DETERMINANT_THRESHOLD`.

Benchmarks remain the same as PR description after these changes.

Also cross product operation has been added for Vec3Fxx

**Comment by Nyx128** [comment]
@cjpurackal 
let me know which changes should i make:
- i will of course update any docs as needed
- change function names and return types as needed
- any other change i should have considered

**Comment by cjpurackal** [comment]
Homography computation doesn't belong in `kornia-algebra`, it should be in `kornia-3d` itself. The algebra crate is reserved for core math.

**Comment by Nyx128** [comment]
> Homography computation doesn't belong in `kornia-algebra`, it should be in `kornia-3d` itself. The algebra crate is reserved for core math. 

Alright, then I will move it to homography, then I suppose you will also want me to provide an array slice interface instead of glam types I suppose.

**Comment by Nyx128** [comment]
@cjpurackal i've completely moved homography operations to kornia-3d as suggested and the performance improvements also stay due to static array implementation. All tests have been ported over as well. Let me know if there is more to this refactor or if you would like any changes

**Comment by Nyx128** [comment]
@cjpurackal can this be reviewed?

**Comment by sidd-27** [comment]
## Overall

Good refactor — consolidates duplicate homography code from `kornia-apriltag` into `kornia-3d`, replaces raw `[f64; 2]`/`[f64; 3]` arrays with proper `Vec2F32`/`Vec2F64`/`Vec3F64` types, and swaps `faer` SVD for `nalgebra`. The direction is sound, but there are several issues worth addressing.

## Issues

**1. New `nalgebra` dependency is heavyweight**

The PR replaces `faer` with `nalgebra` (`SMatrix`, `SVector`, `.svd()`, `.lu()`). Was this discussed? `nalgebra` is a large dependency. If the project was intentionally using `faer` for linear algebra, this swap deserves justification — especially since only small fixed-size matrices are involved. The existing `kornia_algebra::linalg::svd::svd3_f64` is already used for `decompose_homography`; could a similar hand-rolled or `faer`-based approach work for the 8×8/9×9 cases too?

**2. `homography_4pt2d_gauss_elim_f32` — magic index arithmetic is fragile**

The back-substitution results are read via hardcoded flat-array indices (`a[8]`, `a[35]`, `a[62]`, etc.). This is carried over from the old `homography_compute`, but now that you're cleaning things up, consider using a 2D indexing helper or at least adding a comment explaining the index formula (`row * 9 + 8` for the solution column). A single off-by-one here would silently produce wrong homographies.

**3. Determinant threshold inconsistency**

`DETERMINANT_THRESHOLD` is `1e-8` (f64), but in `homography_4pt2d_gauss_elim_f32` it's cast as `DETERMINANT_THRESHOLD as f32`. Meanwhile the Gaussian elimination uses a separate `EPSILON: f32 = 1e-10` for pivot checks. These thresholds serve different purposes — document why they differ, or unify them.

**4. The 9×9 zero-padded matrix in `homography_4pt2d_svd_f64` is suspicious**

You construct a `SMatrix::<f64, 9, 9>` but only fill 8 rows, leaving the 9th row as zeros, explicitly to get a full 9×9 V matrix. This works because the zero row guarantees the 9th singular value is zero, but it's a non-obvious trick. A comment explaining this would help, and it would be more natural to use the 8×9 matrix and take the last column of V directly (nalgebra's SVD supports non-square matrices).

**5. Missing `dot` method addition in `vec.rs`**

The `homography_4pt3d_lu_f64` cheirality checks call `p01.dot(x1[2])`, but the diff only shows `cross` being added to `Vec3F64`/`Vec3F32`/`Vec3AF32`. If `dot` is already available via `glam`'s `Deref` or a trait impl, fine — but this isn't visible in the diff. Worth confirming it compiles.

**6. Error type changed — potential breaking change**

`HomographyError` gained `Clone, Copy, PartialEq`. That's fine, but the function signatures changed from `&mut [[f64; 3]; 3]` out-params to returning `Result<Mat3F64, _>`. If these are public APIs, this is a breaking change that should be called out in the PR description.

**7. `homography_4pt3d_lu_f64` — `h33 = 1.0` assumption**

The LU solver solves for 8 unknowns and hardcodes `h33 = 1.0`. This is correct for the 3D case when `x1[i].z` and `x2[i].z` are non-zero, but will silently produce garbage if points have `z ≈ 0`. A debug assertion or doc note would help.

**8. Test coverage gap**

There's no test for the degenerate/singular case (e.g., collinear points) to verify that `SingularMatrix` is actually returned. The old code also lacked this, but since you're touching all the tests anyway, it's a good time to add it.

## Minor Nits

- The `use nalgebra::{SMatrix, SVector};` import sits between the error enum and the constant — consider grouping imports at the top of the file.
- `homography_project` and `update_homographies` got doc examples added in `quad.rs` — nice.
- The `Cargo.toml` change adding `kornia-3d` as a dependency of `kornia-apriltag` is the correct way to express this consolidation.

## Summary

The core idea (deduplicate, use proper types, return values instead of out-params) is good. Main concerns are the `nalgebra` dependency addition without rationale, the fragile magic indices in the Gauss elimination path, and missing degenerate-case tests.

**Comment by Nyx128** [comment]
@sidd-27 I can make all the improvements mentioned but using `faer` instead of `nalgebra` does make performance regress on benchmarks like `ransac_homography`(back to their original numbers). That was my sole reason for keeping it in the first place. Even so if it happens to be worth that, I can switch back to `faer`. Except this one problem, all the other improvements can be made.

Initially i had put homography inside `kornia-algebra` which then got moved to `kornia-3d`. As such I experimented and the benchmarks stayed, but since SMatrix had to be used for that, i kept it in.

I'll let @cjpurackal decide if nalgebra should be used or not.

File: crates/kornia-3d/src/pose/homography.rs [hunk 6]
```diff
@@ -1,154 +1,297 @@
...
+
+/// Threshold for determinant check to avoid singular matrices.
+const DETERMINANT_THRESHOLD: f64 = 1e-8;
+
+/// Computes the homography matrix from four 2D point correspondences using Gaussian elimination.
 ///
-/// * `x1` - The source 2d points with shape (4, 2).
-/// * `x2` - The destination 2d points with shape (4, 2).
-/// * `homo` - The output homography matrix from src to dst with shape (3, 3).
-pub fn homography_4pt2d(
```
Comment (sidd-27) [inline crates/kornia-3d/src/pose/homography.rs:22]: removing this makes this a breaking api change right?
